sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -1,112 +1,178 @@
+    val ethTx = new ListBuffer[EthereumTransaction]
+    block.transactions.foreach(transaction =>
+      if (transaction.isInstanceOf[EthereumTransaction]) ethTx.append(transaction.asInstanceOf[EthereumTransaction]))
+    ethTx
+  }
+
+  def getBlockById(blockId: String) = nodeView.history.getBlockById(blockId).get()
+
+  @RpcMethod("eth_getBlockByHash") def getBlockByHash(tag: Quantity, hydratedTx: Boolean): EthereumBlock = {
+    val txHashes: ListBuffer[String] = ListBuffer()
+    val txViews: ListBuffer[EthereumTransactionView] = ListBuffer()
+
+    val txs = getTransactionsFromBlock(getBlockById(Numeric.cleanHexPrefix(tag.getValue)))
+    txs.foreach(transaction => {
+      txViews.append(new EthereumTransactionView(transaction))
+      txHashes.append(Numeric.prependHexPrefix(transaction.id))
+    })
+    new EthereumBlock(Numeric.prependHexPrefix(Integer.toHexString(nodeView.history.getBlockHeight(Numeric.cleanHexPrefix(tag.getValue)).get())), Numeric.prependHexPrefix(tag.getValue), txHashes.toList.asJava, txViews.toList.asJava, hydratedTx, nodeView.history.getBlockById(Numeric.cleanHexPrefix(tag.getValue)).get())
+  }
+
+  @RpcMethod("eth_call") def call(params: TransactionArgs, tag: Quantity): Data = {
+    val stateDbRoot = getStateDbRootFromBlockById(getBlockIdByTag(tag.getValue))
+    if (stateDbRoot == null) return new Data(null)
+    val result = Evm.Apply(
+      stateDbRoot.getStateDbHandle,
+      params.getFrom,
+      if (params.to == null) null else params.to.toBytes,
+      params.value,
+      params.getData,
+      params.gas,
+      params.gasPrice
+    )
+
+    if (result.evmError.nonEmpty) {
+      throw new RpcException(
+        new RpcError(
+          RpcCode.ExecutionError.getCode,
+          result.evmError,
+          Numeric.toHexString(result.returnData)
+        )
+      )
     }
-    return new Data(null)
+    new Data(result.returnData)
   }
 
   @RpcMethod("eth_blockNumber") def blockNumber = new Quantity(Numeric.toHexStringWithPrefix(BigInteger.valueOf(getBlockHeight)))
 
+  private def getBlockHeight = nodeView.history.getCurrentHeight
+
+  private def getStateDbRootFromBlockById(blockId: String) : AccountStateView = {
+    val block = nodeView.history.getBlockById(blockId)
+    if (block.isEmpty) return null
+    nodeView.state.getStateDbViewFromRoot(block.get().header.stateRoot)
+  }
+
   @RpcMethod("eth_chainId") def chainId = new Quantity(Numeric.toHexStringWithPrefix(BigInteger.valueOf(networkParams.chainId)))
 
   @RpcMethod("eth_getBalance") def GetBalance(address: String, blockNumberOrTag: Quantity) = new Quantity(getBalance(address, blockNumberOrTag))
 
+  private def getBlockIdByTag(tag: String): String = {
+    var blockId: java.util.Optional[String] = null
+    val history = nodeView.history
+    tag match {
+      case "earliest" => blockId = nodeView.history.getBlockIdByHeight(1)
+      //case "finalized" => //-39001: Unkown block
+      //case "safe" => //-39001: Unkown block
```

Comment:
do we need to support it?
consider that we use this function for blocks/balances/tx count.

Reply:
I just added it